lucet-runtime/lucet-runtime-internals/src/context/mod.rs

```diff
@@ -281,6 +324,41 @@ impl Context {
     /// );
     /// assert!(res.is_ok());
     /// ```
+    ///
+    /// # Implementation details
+    ///
+    /// This prepares a stack for the child context structured as follows, assuming an 0x1000 byte
+    /// stack:
+    /// ```text
+    /// 0x1000: +-------------------------+
+    /// 0x0ff8: | &child                  |
+    /// 0x0ff0: | &parent                 | <-- `backstop_args`, which is stored to `rbp`.
+    /// 0x0fe8: | NULL                    | // Null added if necessary for alignment.
+    /// 0x0fe0: | spilled_arg_1           | // Guest arguments follow.
+    /// 0x0fd8: | spilled_arg_2           |
+    /// 0x0fd0: ~ spilled_arg_3           ~ // The three arguments here are just for show.
+    /// 0x0fc8: | lucet_context_backstop  | <-- This forms an ABI-matching call frame for fptr.
+    /// 0x0fc0: | fptr                    | <-- The actual guest code we want to run.
+    /// 0x0fb8: | lucet_context_bootstrap | <-- The guest stack pointer starts here.
+    /// 0x0fb0: |                         |
+    /// 0x0XXX: ~                         ~ // Rest of the stack needs no preparation.
+    /// 0x0000: |                         |
+    ///         +-------------------------+
+    /// ```
+    ///
+    /// This packing of data on the stack is interwoven with noteworthy constraints on what the
```

Comment:
Explicitly writing out the constraints this deals with made an interesting option visible:
`lucet_context_bootstrap` can load initial volatile volatile registers (so, arguments from `fptr`) from the stack, with an initial thunk like
```
pop rdi
pop rsi
pop rcx
pop rdx
pop r8
pop r9
ret
```
This frees up all callee-saves for use however we'd want - and means we can keep the parent and child contexts in, for example, `r12` and `r13`. This also avoids forcing the [gymnastics](https://github.com/fastly/lucet/pull/273/files#diff-adb782ea0cfdba94fce93722c7e66267R98) I needed to do to thread the heap address through to r15 in the guest, since r15 becomes available in the initial `Context` as well.

Reply:
How would this interact with yield/resume? Would the new `Context::swap()` still be suitable even if we didn't have a function entrypoint?

Reply:
Ah this wouldn't change what `Context::swap()` works with. As far as the guest call stack would be concerned, nothing changes. These are all caller-save registers, so they're not saved in `swap()` anyway, and there's no difference there. So the only change is that initial argument registers that make it to `fptr` would come from the stack, instead of through callee-saves we set up the initial Context with

Reply:
Got it, that sounds good!